### PR TITLE
Fixed #28378 -- Fixed union(), and difference() when combining with queryset raising EmptyResultSet exception.

### PR DESCRIPTION
**Pull request by felixxm**
https://code.djangoproject.com/ticket/28378

### CODE REVIEW

**Comment by felixxm** [comment]
buildbot, test on oracle.

File: django/db/models/sql/compiler.py [hunk 1]
```diff
@@ -439,6 +443,15 @@ def as_sql(self, with_limits=True, with_col_aliases=False):
...
             else:
+                distinct_fields = self.get_distinct()
+
+                # This must come after 'select', 'ordering', and 'distinct'
+                # (see docstring of get_from_clause() for details).
+                from_, f_params = self.get_from_clause()
+
+                where, w_params = self.compile(self.where) if self.where is not None else ("", [])
+                having, h_params = self.compile(self.having) if self.having is not None else ("", [])
+
```
Comment (charettes) [inline django/db/models/sql/compiler.py]: I haven't checked out the patch locally but were these changes necessary for the patch to work?

Comment (felixxm) [inline django/db/models/sql/compiler.py]: Yes. `EmptyResultSet` exception is raised here, therefore I decided to move these variables inside the `if` clause where they are used.

**Comment by felixxm** [comment]
@charettes Thanks for review. I added release notes.

**Comment by felixxm** [comment]
buildbot, test on oracle.

File: docs/releases/1.11.4.txt [hunk 2]
```diff
@@ -12,3 +12,6 @@ Bugfixes
 * Fixed a regression in 1.11.3 on Python 2 where non-ASCII ``format`` values
   for date/time widgets results in an empty ``value`` in the widget's HTML
   (:ticket:`28355`).
+
+* Fixed ``QuerySet.union()``, and ``difference()`` when combining with queryset
```
Comment (timgraham) [inline docs/releases/1.11.4.txt]: chop comma

File: docs/releases/1.11.4.txt [hunk 3]
```diff
@@ -12,3 +12,6 @@ Bugfixes
...
+  raising ``EmptyResultSet`` exception (:ticket:`28378`).
```
Comment (timgraham) [inline docs/releases/1.11.4.txt]: "exception" seems unneeded

File: docs/releases/1.11.4.txt (hunk 2 above)
Comment (timgraham) [inline docs/releases/1.11.4.txt]: with a queryset

File: django/db/models/sql/compiler.py [hunk 4]
```diff
@@ -399,7 +399,18 @@ def get_combinator_sql(self, combinator, all):
                     raise DatabaseError('LIMIT/OFFSET not allowed in subqueries of compound statements.')
                 if compiler.get_order_by():
                     raise DatabaseError('ORDER BY not allowed in subqueries of compound statements.')
-        parts = (compiler.as_sql() for compiler in compilers)
+        parts = ()
+        for compiler in compilers:
+            try:
+                parts += (compiler.as_sql(),)
+            except EmptyResultSet:
+                # Omit the empty queryset with UNION operator and with
```
Comment (timgraham) [inline django/db/models/sql/compiler.py]: with UNION and with DIFFERENCE if the first queryset is nonempty.

File: django/db/models/sql/compiler.py [hunk 5]
```diff
@@ -422,15 +433,8 @@ def as_sql(self, with_limits=True, with_col_aliases=False):
         refcounts_before = self.query.alias_refcount.copy()
         try:
             extra_select, order_by, group_by = self.pre_sql_setup()
-            distinct_fields = self.get_distinct()
-
-            # This must come after 'select', 'ordering', and 'distinct' -- see
-            # docstring of get_from_clause() for details.
-            from_, f_params = self.get_from_clause()
```
Comment (timgraham) [inline django/db/models/sql/compiler.py]: I'd chop the blank lines around `for_update_part`.

File: django/db/models/sql/compiler.py [hunk 6]
```diff
@@ -439,6 +443,15 @@ def as_sql(self, with_limits=True, with_col_aliases=False):
                     raise NotSupportedError('{} is not supported on this database backend.'.format(combinator))
                 result, params = self.get_combinator_sql(combinator, self.query.combinator_all)
             else:
+                distinct_fields = self.get_distinct()
+
```
Comment (timgraham) [inline django/db/models/sql/compiler.py]: I'd omit the blank lines.

**Comment by felixxm** [comment]
@timgraham Thanks for review.

**Comment by felixxm** [comment]
Merged in 9bca0d0b38d941fe7f3842cb2259d018823ed25e and ca74e563500e291480f1976b58fcd34aac768dca.
